lib/compiler/src/beam_ssa_opt.erl

```diff
@@ -3711,6 +3712,89 @@ build_bs_ensure_match(L, {_,Size,Unit}, Count0, Blocks0) ->
 
     {Blocks,Count}.
 
+%%%
+%%% Change the `reuse` hint to `copy` when it is highly probable that
+%%% reuse will not happen.
+%%%
+
+ssa_opt_no_reuse({#opt_st{ssa=Linear0}=St, FuncDb}) when is_list(Linear0) ->
+    New = sets:new([{version,2}]),
+    Linear = ssa_opt_no_reuse_blks(Linear0, New),
+    {St#opt_st{ssa=Linear}, FuncDb}.
+
+ssa_opt_no_reuse_blks([{L,#b_blk{is=Is0}=Blk0}|Bs], New0) ->
+    {Is,New} = ssa_opt_no_reuse_is(Is0, New0, []),
+    Blk = Blk0#b_blk{is=Is},
+    [{L,Blk}|ssa_opt_no_reuse_blks(Bs, New)];
+ssa_opt_no_reuse_blks([], _) ->
+    [].
+
+ssa_opt_no_reuse_is([#b_set{op=update_record,args=Args}=I0|Is], New, Acc) ->
+    [_,_,_|Updates] = Args,
+    case cannot_reuse(Updates, New) of
+        true ->
+            I = I0#b_set{args=[#b_literal{val=copy}|tl(Args)]},
+            ssa_opt_no_reuse_is(Is, New, [I|Acc]);
+        false ->
+            ssa_opt_no_reuse_is(Is, New, [I0|Acc])
+    end;
+ssa_opt_no_reuse_is([#b_set{dst=Dst}=I|Is], New0, Acc) ->
+    case inhibits_reuse(I, New0) of
+        true ->
+            New = sets:add_element(Dst, New0),
+            ssa_opt_no_reuse_is(Is, New, [I|Acc]);
+        false ->
+            ssa_opt_no_reuse_is(Is, New0, [I|Acc])
+    end;
+ssa_opt_no_reuse_is([], New, Acc) ->
+    {reverse(Acc),New}.
+
+inhibits_reuse(#b_set{op=phi,args=Args}, New) ->
+    foldl(fun({Value,_}, Bool) ->
+                  sets:is_element(Value, New) andalso Bool
```

Comment:
I don't know what I was thinking when I wrote this line:

```suggestion
                  Bool andalso sets:is_element(Value, New)
```

Reply:
`lists:all/1` maybe?

Reply:
> I don't know what I was thinking when I wrote this line

I had made it short-circuiting in the version which did module global analysis. FYI @jhogberg, extending this analysis to be module global increases the time spent  on this pass by a factor of between 4 and 5 and only triggers infrequently, and for example, never allows for additional destructive updates. From that @bjorng and I decided that doing strength reduction at the function level is good enough.

Apropos `lists:all/2`, shouldn't we document that it is short-circuiting? I sometimes find myself thinking about things like that, and when it's not documented I tend to ad-hoc it instead of looking into the implementation.